### PR TITLE
Adding Role entity and all related APIs: Exists, Members and Memberships

### PR DESCRIPTION
**Pull request by acher82**
**What does this pull request do?**
Exposes Role related APIs: [Exists](https://conjur.docs.apiary.io/#reference/role/exists/determine-whether-a-role-exists), [Members](https://conjur.docs.apiary.io/#reference/role/list-members/lists-the-roles-that-have-been-the-recipient-of-a-role-grant) and [Memberships](https://conjur.docs.apiary.io/#reference/role/list-memberships/list-the-roles-a-role-is-a-member-of).
**What background context can you provide?**
This functionality is nesessary in Vault-Conjur Synchroniser.
We want to be sure that users and groups that we created related each other as expected.
**Where should the reviewer start?**
conjur-api/Role.cs

### CODE REVIEW

File: conjur-api/ResourceMetadata.cs [hunk 1]
```diff
@@ -59,4 +59,16 @@ public class Annoatation
         public string UpdatedAt { get; private set; }
     }
 
+    [DataContract]
+    public class RoleMember
+    {
+        [DataMember(Name = "admin_option")]
+        public bool Admin { get; set; }
```
Comment (DvirCyberArk) [inline conjur-api/ResourceMetadata.cs]: Can we keep this name aligned with other members (admin != admin_options)? e.g UpdatedAt.

File: conjur-api/Role.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+namespace Conjur
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Net;
+
+    /// <summary>
+    /// Conjur Role reference.
+    /// </summary>
+    /// A role is an actor in the system, in the classical sense of role-based access control. Roles are the entities which receive permission grants.
```
Comment (DvirCyberArk) [inline conjur-api/Role.cs]: Should we add "A role..." to summary tag? and not as standalone documentation?

File: conjur-api/Role.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Conjur.Role"/> class.
+        /// </summary>
+        /// <param name="client">Conjur client to use to connect.</param>
+        /// <param name="kind">Kind of the role, for example 'group' or 'layer'.</param>
+        /// <param name="name">The Role name.</param>
+        internal Role(Client client, string kind, string name)
+            : base(client, kind, name)
+        {
+            this.path = $"authz/{client.GetAccountName()}/roles/{WebUtility.UrlEncode(kind)}/{WebUtility.UrlEncode(name)}";
```
Comment (DvirCyberArk) [inline conjur-api/Role.cs]: Maybe accountName should be encoded also?

File: conjur-api/Role.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        /// <summary>
+        /// Check for the existence of a role.
+        /// </summary>
+        /// <returns>True if role exists otherwice false.</returns>
+        /// Only roles that you have read permission on will be searched.
```
Comment (DvirCyberArk) [inline conjur-api/Role.cs]: We want to leave standalone documentation lines? or add them to tags: summary / returns

File: conjur-api/Role.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            WebRequest webRequest = this.Client.AuthenticatedRequest(this.path);
+            webRequest.Method = WebRequestMethods.Http.Head;
+            try
+            {
+                webRequest.GetResponse().Close();                
+            }
+            catch (WebException ex)
+            {
+                HttpWebResponse responce = ex.Response as HttpWebResponse;
```
Comment (DvirCyberArk) [inline conjur-api/Role.cs]: TYPO - responce => response

File: test/RoleTest.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        public RoleTest()
+        {
+            Client.Authenticator = new MockAuthenticator();
+        }
+
+        [Test]
+        public void TestRoleDoesExist()
+        {
+            ClearMocker();
+            Mocker.Mock(new Uri($"{baseUrl}/groupName"), "");
```
Comment (DvirCyberArk) [inline test/RoleTest.cs]: Refactor it as constant? and return String.Empty?

File: test/RoleTest.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            Assert.IsTrue(Client.Role("group", "groupName").Exists(), "Group 'groupName' expected to be exist");
```
Comment (DvirCyberArk) [inline test/RoleTest.cs]: Use existing constant of kind group?

File: test/RoleTest.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            ClearMocker();
+            Mocker.Mock(new Uri($"{baseUrl}/groupName?members"), GenerateMembersData(10));
+            VerifyRoleInfo(Client.Role("group", "groupName").ListMembers().GetEnumerator(), 10);
+        }
+
+        [Test]
+        public void TestRoleMemberships()
+        {
+            ClearMocker();
+            Mocker.Mock(new Uri($"{baseUrl}/groupName?all"), "[ \"role0\", \"role1\" ]");
```
Comment (DvirCyberArk) [inline test/RoleTest.cs]: Extract it to methods like GenerateMemberData?

File: test/RoleTest.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            string[] roles = Client.Role("group", "groupName").ListMemberships().ToArray();
+            Assert.AreEqual("role0", roles[0]);
+            Assert.AreEqual("role1", roles[1]);
+        }
+
+        private void VerifyRoleInfo(IEnumerator<RoleMember> members, int expectedNum)
```
Comment (DvirCyberArk) [inline test/RoleTest.cs]: Since we are not manipulating theMember{id}, it seems like we check only the mechanism of Mocker.mock and Verify - Should we keep testing those section?

**Comment by DvirCyberArk** [comment]
Hi @acher82 overall looks good, added some comments - should we use a template for a poll request description? 

e.g: https://github.com/cyberark/conjur/pull/353

TL;DR: 
What does this pull request do?
What background context can you provide?
Where should the reviewer start?
How should this be manually tested?
Link to build in Jenkins (if appropriate) - do we have custom build?
Questions:

Probably we have this template documented somewhere... :) @ryanprior can you direct us?

File: conjur-api/Role.cs (hunk 3 above)
Comment (acher82) [reply conjur-api/Role.cs]: You are right. Fixed also in ListResources method.

File: conjur-api/Role.cs (hunk 4 above)
Comment (acher82) [reply conjur-api/Role.cs]: I'll wrap it with remarks tag (see https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/remarks)

File: conjur-api/Role.cs (hunk 5 above)
Comment (acher82) [reply conjur-api/Role.cs]: Fixed

File: test/RoleTest.cs (hunk 6 above)
Comment (acher82) [reply test/RoleTest.cs]: Fixed string.empty. Dealing with consts in UT feels me unnecessary.

File: test/RoleTest.cs (hunk 8 above)
Comment (acher82) [reply test/RoleTest.cs]: I think it's pretty simple for dedicated method.

File: test/RoleTest.cs (hunk 9 above)
Comment (acher82) [reply test/RoleTest.cs]: We test that we get exactly what we expect to get.
How do you see this unit test?

**Comment by acher82** [comment]
Thank for CR I've added answers for relevant questions from the list
